subsys/bluetooth/lib/encrypted_ad_data/encrypted_ad_data.c

```diff
@@ -0,0 +1,108 @@
+/* Copyright (c) 2023 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/bluetooth/encrypted_ad_data.h>
+
+#include <stdint.h>
+#include <string.h>
+
+#include <zephyr/logging/log.h>
+#include <zephyr/bluetooth/crypto.h>
+
+LOG_MODULE_REGISTER(bt_encrypted_ad_data, CONFIG_BT_ENCRYPTED_AD_DATA_LOG_LEVEL);
+
+static int bt_ead_generate_randomizer(uint8_t *randomizer)
+{
+	int err;
+
+	err = bt_rand(randomizer, BT_EAD_RANDOMIZER_SIZE);
+
+	if (err != 0) {
+		return err;
+	}
+
+	/* From Supplement to the Bluetooth Core Specification v11, Part A 1.23.3: The directionBit
+	 * of the CCM nonce shall be set to the most significant bit of the Randomizer field.
+	 */
+	randomizer[4] |= 1 << BT_EAD_RANDOMIZER_DIRECTION_BIT;
+
+	return 0;
+}
+
+static int bt_ead_generate_nonce(const uint8_t *iv, uint8_t *nonce, const uint8_t *randomizer)
+{
+	int err;
+	uint8_t new_randomizer[BT_EAD_RANDOMIZER_SIZE];
+
+	if (randomizer == NULL) {
+		err = bt_ead_generate_randomizer(new_randomizer);
+
+		if (err != 0) {
+			LOG_ERR("Failed to generate Randomizer");
+			return err;
+		}
+
+		randomizer = new_randomizer;
+	}
+
+	memcpy(&nonce[0], randomizer, BT_EAD_RANDOMIZER_SIZE);
+	memcpy(&nonce[BT_EAD_RANDOMIZER_SIZE], iv, BT_EAD_IV_SIZE);
+
+	return 0;
+}
+
+int bt_ead_encrypt_ad_data(const uint8_t *session_key, const uint8_t *iv, const uint8_t *payload,
+			   size_t payload_size, uint8_t *encrypted_ad_data)
+{
+	int err;
+	uint8_t nonce[BT_EAD_NONCE_SIZE];
+	int ead_size = payload_size + BT_EAD_RANDOMIZER_SIZE + BT_EAD_AAD_SIZE + BT_EAD_MIC_SIZE;
+
+	bt_ead_generate_nonce(iv, nonce, NULL);
+
+	memcpy(encrypted_ad_data, nonce, BT_EAD_RANDOMIZER_SIZE);
+	memcpy(encrypted_ad_data + BT_EAD_RANDOMIZER_SIZE, bt_ead_aad, BT_EAD_AAD_SIZE);
+
+	err = bt_ccm_encrypt(session_key, nonce, payload, payload_size, bt_ead_aad, BT_EAD_AAD_SIZE,
+			     encrypted_ad_data + BT_EAD_RANDOMIZER_SIZE + BT_EAD_AAD_SIZE,
+			     BT_EAD_MIC_SIZE);
+
+	LOG_HEXDUMP_DBG(encrypted_ad_data, ead_size, "Encrypted Data: ");
+
+	if (err != 0) {
+		LOG_ERR("Failed to encrypt the payload (bt_ccm_encrypt err %d)", err);
+		return -EIO;
+	}
+
+	return ead_size;
+}
+
+int bt_ead_decrypt_ad_data(const uint8_t *session_key, const uint8_t *iv,
+			   const uint8_t *encrypted_ad_data, size_t encrypted_ad_data_size,
+			   uint8_t *payload)
+{
+	int err;
+	uint8_t nonce[BT_EAD_NONCE_SIZE];
+	const uint8_t *encrypted_payload =
+		encrypted_ad_data + BT_EAD_RANDOMIZER_SIZE + BT_EAD_AAD_SIZE;
+	int encrypted_payload_size =
+		(encrypted_ad_data_size - BT_EAD_RANDOMIZER_SIZE) - BT_EAD_AAD_SIZE;
+	int payload_size = encrypted_payload_size - BT_EAD_MIC_SIZE;
```

Comment:
Hmm, this may be tricky if we want to return the payload_size as an `int` as well. 

I assume that we'll never see an overflow, but still. Perhaps `ssize_t`?